app/controllers/auth_controller.rb

```diff
@@ -31,6 +31,15 @@ def fetch_return_to_path
   def do_user_auth(auth)
     sign_out
     user = User.from_oauth_hash(auth)
+    unless ENV["NO_WELCOME_EMAIL"]
+      send_welcome_mail(user)
+    end
     sign_in(user)
   end
+
+  def send_welcome_mail(user)
+    if (Time.current - user.created_at) < 2.seconds
+      WelcomeMailer.welcome_notification(user).deliver_later
```

Comment:
there wasn't anything client-specific about a "welcome to the product" message so my initial thought was to skip the extra level of indirection. I'm happy to make it a Dispatcher method if you think consistency is most important.